firebaseai/LiveAudioExample/Audio/AudioController.swift

```
@@ -0,0 +1,256 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import AVFoundation
+
+/// Controls audio playback and recording.
+actor AudioController {
+  /// Data processed from the microphone.
+  private let microphoneData: AsyncStream<AVAudioPCMBuffer>
+  private let microphoneDataQueue: AsyncStream<AVAudioPCMBuffer>.Continuation
+  private var audioPlayer: AudioPlayer?
+  private var audioEngine: AVAudioEngine?
+  private var microphone: Microphone?
+  private var listenTask: Task<Void, Never>?
+  private var routeTask: Task<Void, Never>?
+
+  /// Port types that are considered "headphones" for our use-case.
+  ///
+  /// More specifically, airpods are considered bluetooth ports instead of headphones, so
+  /// this array is necessary.
+  private let headphonePortTypes: [AVAudioSession.Port] = [
+    .headphones,
+    .bluetoothA2DP,
+    .bluetoothLE,
+    .bluetoothHFP,
+  ]
+
+  private let modelInputFormat: AVAudioFormat
+  private let modelOutputFormat: AVAudioFormat
+
+  private var stopped = false
+
+  public init() async throws {
+    let session = AVAudioSession.sharedInstance()
+    try session.setCategory(
+      .playAndRecord,
+      mode: .voiceChat,
+      options: [.defaultToSpeaker, .allowBluetooth, .duckOthers,
+                .interruptSpokenAudioAndMixWithOthers, .allowBluetoothA2DP]
+    )
+    try session.setPreferredIOBufferDuration(0.01)
+    try session.setActive(true)
+
+    guard let modelInputFormat = AVAudioFormat(
+      commonFormat: .pcmFormatInt16,
+      sampleRate: 16000,
+      channels: 1,
+      interleaved: false
+    ) else {
+      fatalError("Failed to create model input format")
+    }
+
+    guard let modelOutputFormat = AVAudioFormat(
+      commonFormat: .pcmFormatInt16,
+      sampleRate: 24000,
+      channels: 1,
+      interleaved: true
+    ) else {
+      fatalError("Failed to create model output format")
```

Same here - it'd be better to raise an error and handle it in the UI.